src/merkle/tiered_smt/mod.rs

```diff
@@ -166,6 +166,34 @@ impl TieredSmt {
         TieredSmtProof::new(path, entries).expect("Bug detected, TSMT produced invalid proof")
     }
 
+    /// Provides the data structures necessary to perform a sequence of key inserts in the Miden VM.
+    /// Specifically, the partial Merkle tree is used to populate the advice provider's Merkle
+    /// store, while `advice_map` is used to populate the advice map.
+    pub fn get_insert_witness<'a>(
+        &'a self,
+        keys_to_insert: impl IntoIterator<Item = &'a RpoDigest>,
+    ) -> Result<TieredSmtInsertWitness, MerkleError> {
+        let mut partial_mt = PartialMerkleTree::default();
+        let mut advice_map = BTreeMap::new();
+
+        for key in keys_to_insert {
+            let (leaf_index, leaf_node_exists) = self.nodes.get_leaf_index(key);
+            let path = self.get_path(leaf_index.into())?;
+            let leaf_node = self.get_node(leaf_index.into())?;
+
+            partial_mt.add_path(leaf_index.value(), leaf_node, path)?;
+
+            // Update advice_map with all key/values who share the same node as current `key`
+            if leaf_node_exists {
+                let kvs_at_leaf: Vec<_> = self.values.iter_leaf(leaf_index).cloned().collect();
```

Comment:
I think we should be able to replace this with something like:
```Rust
let kvs_at_leaf = if leaf_index.value() == Self::MAX_DEPTh {
    self.values.get_all(get_key_prefix(key)).expect("no leaf")
} else {
    vec![self.values.get_first(index_to_prefix(leaf_index)).expect("no leaf")]
}
```
Basically, if we know we are at the bottom tier, we the whole bottom leaf. If not, then we get the first leaf which matches the prefix of the key at the leaf's depth.

If the above works, we won't need `iter_leaf()` method on `ValueStore`.

Reply:
Ah yes, that works. I now realize I thought I needed `iter_leaf()` before I used the `leaf_node_exists` flag (to prevent `get_first()` from grabbing something from another leaf).

`iter_leaf()` removed